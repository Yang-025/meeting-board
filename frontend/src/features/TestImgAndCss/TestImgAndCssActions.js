// import { createAction } from 'redux-actions';
import Resource from '~~apis/resource';


export const actionTypes = {
  A_SIMPLE_ACTION: 'A_SIMPLE_ACTION',
  ANOTHER_SIMPLE_ACTION: 'ANOTHER_SIMPLE_ACTION',
  SIMPLE_SAGA_SUCCESS: 'SIMPLE_SAGA_SUCCESS',
  SIMPLE_SAGA_FAILURE: 'SIMPLE_SAGA_FAILURE',
  ERROR: 'ERROR',
  OBSERVABLE_PING: 'OBSERVABLE_PING',
  OBSERVABLE_PONG: 'OBSERVABLE_PONG'
};


function callApiTestSuccess(response) {
  // console.log('callApiTestSuccess>>', response);
  return {
    type: actionTypes.ANOTHER_SIMPLE_ACTION,
    api: response.data
  };
}

function callApiTestFailed(error) {
  // console.log('callApiTestFailed>>', error.message);
  return {
    type: actionTypes.ERROR,
    payload: error.message
  };
}

// 好像一定要return axios({}) 才有辦法做測試
export function anotherSimpleAction() {
  return ((dispatch) => {
    const data = {
      title: 'foo222',
      body: 'bar222',
      userId: 1
    };
    return Resource.UserResource.create(data)
    // return User.get()
      .then(response => dispatch(callApiTestSuccess(response)))
      .catch(error => dispatch(callApiTestFailed(error)));
  });
}

export function anotherSimpleAction2() {
  return ((dispatch) => {
    return Resource.UserResource.get2()
      .then(response => dispatch(callApiTestSuccess(response)))
      .catch(error => dispatch(callApiTestFailed(error)));
  });
}

export function simpleAction() {
  return {
    type: actionTypes.A_SIMPLE_ACTION,
    msg: 'KERKERKER'
  };
}


export function ping() {
  return {
    type: actionTypes.OBSERVABLE_PING
  };
}

export function pong() {
  return {
    type: actionTypes.OBSERVABLE_PONG
  };
}
