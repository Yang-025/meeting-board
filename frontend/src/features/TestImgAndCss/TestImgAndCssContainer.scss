.normal {
  border: 10px solid goldenrod;
  /* 測試 prefix */
  display: flex;
  background: linear-gradient(red, green);
}

.hovered {
  border: 10px solid forestgreen;
  /* 測試 prefix */
  display: flex;
  background: linear-gradient(red, green);
}

// :global可以做全局改變
:global .normal-css {
  border: 10px solid teal;
  /* 測試 prefix */
  display: flex;
  background: linear-gradient(red, green);

  &:hover {
    border: 10px solid peru;
  }
  
}