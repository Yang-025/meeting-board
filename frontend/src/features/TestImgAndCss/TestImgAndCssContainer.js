import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import styled from 'styled-components';
import small from '~~assets/images/small.jpg';
import big from '~~assets/images/big.jpg';
import { simpleAction, anotherSimpleAction, anotherSimpleAction2, ping } from './TestImgAndCssActions';

import './TestImgAndCssContainer.scss';


const IMG = styled.img`
  border: 10px solid ${props => (props.status === 'normal' ? 'goldenrod' : 'forestgreen')};
  /* 測試 prefix */
  display: flex;
  background: linear-gradient(red, green);
`;


const STATUS = {
  HOVERED: 'hovered',
  NORMAL: 'normal'
};
@connect(
  state => ({
    msg: state.simple.msg,
    status: state.simple.status,
    isPinging: state.simple.isPinging
  }),
  dispatch => ({
    // actions: bindActionCreators(TestImgAndCssActions, dispatch),
    actions: bindActionCreators({ simpleAction, anotherSimpleAction, anotherSimpleAction2, ping }, dispatch)
  }),
)
class TestImgAndCss extends Component {
  constructor(props) {
    super(props);
    this.state = {
      class: STATUS.NORMAL
    };
  }

  componentDidMount() {
    console.log('TestImgAndCss componentDidMount!!');
    console.log('[TestImgAndCss要近來]');
    this.props.actions.anotherSimpleAction();
    // this.props.actions.anotherSimpleAction2();
  }

  componentWillUnmount() {
    console.log('[TestImgAndCss要離開]');
  }

  onMouseEnter = () => {
    this.setState({ class: STATUS.HOVERED });
  }


  onMouseLeave = () => {
    this.setState({ class: STATUS.NORMAL });
  }


  render() {
    return (
      <div>
        <button onClick={this.props.actions.simpleAction}>Click Me</button>
        <span>{this.props.msg} {this.props.status}</span>
        <button onClick={this.props.actions.ping}>Rx-Observable Test</button>
        <span>{this.props.isPinging.toString()}</span>
        <IMG
          className={this.state.class}
          status={this.state.class}
          src={big}
          alt="big"
          onMouseEnter={this.onMouseEnter}
          onMouseLeave={this.onMouseLeave}
        />
        <img className="normal-css" src={small} alt="small" />
      </div>
    );
  }
}

export default TestImgAndCss;


// props驗證
TestImgAndCss.WrappedComponent.propTypes = {
  status: PropTypes.string.isRequired,
  msg: PropTypes.string.isRequired,
  actions: PropTypes.shape({
    simpleAction: PropTypes.func.isRequired,
    anotherSimpleAction: PropTypes.func.isRequired,
    anotherSimpleAction2: PropTypes.func.isRequired
  }).isRequired
};

// 預設props
// TestImgAndCss.WrappedComponent.defaultProps = {};
