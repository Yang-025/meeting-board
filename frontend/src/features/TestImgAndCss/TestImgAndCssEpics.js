// test dispatch
// import 'rxjs';
import { ofType } from 'redux-observable';
import { mapTo, delay } from 'rxjs/operators';
import { pong, actionTypes } from './TestImgAndCssActions';


export const pingEpic = action$ => {
  return action$.pipe(
    ofType(actionTypes.OBSERVABLE_PING),
    delay(3000),
    mapTo(pong())
  );
};

export default [
  pingEpic
];
