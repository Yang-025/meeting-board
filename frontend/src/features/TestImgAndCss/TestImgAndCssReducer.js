import { handleActions } from 'redux-actions';
import { actionTypes } from './TestImgAndCssActions';

const initialState = {
  count: 0,
  msg: 'Hi',
  status: 'wait...',
  isPinging: false,
};

export default handleActions({
  [actionTypes.A_SIMPLE_ACTION]: (state, payload) => {
    const { msg } = payload;
    return {
      ...state,
      msg
    };
  },
  [actionTypes.SIMPLE_SAGA_SUCCESS]: (state, payload) => {
    const { status } = payload;
    return {
      ...state,
      status
    };
  },
  [actionTypes.SIMPLE_SAGA_FAILURE]: (state, payload) => {
    const { status } = payload;
    return {
      ...state,
      status
    };
  },
  [actionTypes.OBSERVABLE_PING]: (state) => {
    return {
      ...state,
      isPinging: 'true'
    };
  },
  [actionTypes.OBSERVABLE_PONG]: (state) => {
    return {
      ...state,
      isPinging: false
    };
  }
}, initialState);
