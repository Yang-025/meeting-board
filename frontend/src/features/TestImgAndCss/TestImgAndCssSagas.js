import { put, takeEvery } from 'redux-saga/effects';
import { actionTypes } from './TestImgAndCssActions';


function* simpleSaga() {
  try {
    yield put({ type: actionTypes.SIMPLE_SAGA_SUCCESS, status: 'success' });
  } catch (e) {
    yield put({ type: actionTypes.SIMPLE_SAGA_FAILURE, status: e.message });
  }
}

export default function* testImgAndCssSagas() {
  yield [
    takeEvery(actionTypes.A_SIMPLE_ACTION, simpleSaga),
    takeEvery(actionTypes.OBSERVABLE_PING, simpleSaga),
  ];
}
