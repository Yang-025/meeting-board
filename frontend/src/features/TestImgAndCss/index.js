import TestImgAndCssContainer from './TestImgAndCssContainer';

export default TestImgAndCssContainer;
