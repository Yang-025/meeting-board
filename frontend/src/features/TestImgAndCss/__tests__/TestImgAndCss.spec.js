import React from 'react';
import { shallow, configure } from 'enzyme';
import renderer from 'react-test-renderer';
import TestImgAndCss from '~~features/TestImgAndCss';


import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

// jest.mock('react-redux', () => ({ connect: jest.fn() }));

// https://github.com/airbnb/enzyme/issues/472 使用WrappedComponent測試container
describe('<TestImgAndCss> /', () => {
  test('contains <TestImgAndCss/> component', () => {
    // This is where we create a mock state/props
    const props = {
      status: 'Nice post!',
      msg: 'Tiffany Wu',
      isPinging: false,
      actions: {
        simpleAction: jest.fn(() => {}),
        anotherSimpleAction: jest.fn(() => {}),
        anotherSimpleAction2: jest.fn(() => {})
      }
    };
    const wrapper = shallow(<TestImgAndCss.WrappedComponent {...props} />);
    expect(wrapper.find('img')).toHaveLength(1);
  });

  test('renders TestImgAndCss', () => {
    const props = {
      status: 'Nice post!',
      msg: 'Tiffany Wu',
      isPinging: false,
      actions: {
        simpleAction: jest.fn(() => {}),
        anotherSimpleAction: jest.fn(() => {}),
        anotherSimpleAction2: jest.fn(() => {})
      }
    };
    const wrapper = shallow(<TestImgAndCss.WrappedComponent {...props} />);
    expect(wrapper).toHaveLength(1);
  });
});


describe('<TestImgAndCss> 的截圖/', () => {
  test('大圖hover時，class有改變', () => {
    const props = {
      status: 'Nice post!',
      msg: 'Tiffany Wu',
      isPinging: false,
      actions: {
        simpleAction: jest.fn(() => {}),
        anotherSimpleAction: jest.fn(() => {}),
        anotherSimpleAction2: jest.fn(() => {})
      }
    };
    const component = renderer.create(<TestImgAndCss.WrappedComponent {...props} />);
    // 先截一張看看
    let tree = component.toJSON();
    const [bigImg] = tree.children.filter(o => o.type === 'img' && o.props.alt === 'big');

    // console.log('child', tree.children.filter(o => o.type === 'img' && o.props.alt === 'big'));
    // console.log('tree props', tree.props);
    expect(tree).toMatchSnapshot();

    // 呼叫大圖的onMouseEnter()後再截一張
    bigImg.props.onMouseEnter();
    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // 呼叫大圖的onMouseLeave()後再截一張
    bigImg.props.onMouseLeave();
    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
