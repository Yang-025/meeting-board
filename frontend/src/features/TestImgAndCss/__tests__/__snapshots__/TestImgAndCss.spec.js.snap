// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TestImgAndCss> 的截圖/ 大圖hover時，class有改變 1`] = `
<div>
  <button
    onClick={[MockFunction]}
  >
    Click Me
  </button>
  <span>
    Tiffany Wu
     
    Nice post!
  </span>
  <button
    onClick={undefined}
  >
    Rx-Observable Test
  </button>
  <span>
    false
  </span>
  <img
    alt="big"
    className="normal sc-bdVaJa JbXtA"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    src="test-file-stub"
  />
  <img
    alt="small"
    className="normal-css"
    src="test-file-stub"
  />
</div>
`;

exports[`<TestImgAndCss> 的截圖/ 大圖hover時，class有改變 2`] = `
<div>
  <button
    onClick={[MockFunction]}
  >
    Click Me
  </button>
  <span>
    Tiffany Wu
     
    Nice post!
  </span>
  <button
    onClick={undefined}
  >
    Rx-Observable Test
  </button>
  <span>
    false
  </span>
  <img
    alt="big"
    className="hovered sc-bdVaJa biSeJy"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    src="test-file-stub"
  />
  <img
    alt="small"
    className="normal-css"
    src="test-file-stub"
  />
</div>
`;

exports[`<TestImgAndCss> 的截圖/ 大圖hover時，class有改變 3`] = `
<div>
  <button
    onClick={[MockFunction]}
  >
    Click Me
  </button>
  <span>
    Tiffany Wu
     
    Nice post!
  </span>
  <button
    onClick={undefined}
  >
    Rx-Observable Test
  </button>
  <span>
    false
  </span>
  <img
    alt="big"
    className="normal sc-bdVaJa JbXtA"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    src="test-file-stub"
  />
  <img
    alt="small"
    className="normal-css"
    src="test-file-stub"
  />
</div>
`;
